File: interview/Interleaving.java
/*************************************************************************
*  Author: Sriharsha Gangam
*
*  Compilation:  javac Interleaving.java
*  Execution:    java Interleaving
*
*  Interleaving: Takes three strings and tests if the third string is
*  an interleaving of the first two strings.
*
*************************************************************************/



public class Interleaving {
    /*
     * Tests if String z is an interleaving of x and y.
     * A string z interleaves x and y if (1) all the characters of
     * x and y are present in z  (2) all characters of z either belong
     * to x or y and, (3) the order of characters in z follow the
     * same order in their origina strings x and y.
     * Example: "abpcqrd"  is an interleaving of string "abcd" and "pqr"
     *
     * Gloal: Given three strings, the function determines if one is an
     * interleaving of the other two.
     */
    public static boolean interleaving (String x, String y, String z) {
        int xlen = x.length(); int ylen = y.length(); int zlen = z.length();

        boolean [][][] interleavingState = new boolean [xlen + 1][ylen + 1][zlen + 1]; 
        interleavingState[0][0][0] = true;
        interleavingState[1][0][0] = interleavingState[0][1][0] = interleavingState[0][0][1] = interleavingState[1][1][0] = interleavingState[1][1][1] = false;
        

        System.out.println("In Interleaving");
        return true;
    }

    public static void main (String[] args){

        assert interleaving("a","","a");
        assert interleaving("b","","b");
        assert interleaving("a","p","ap");
        assert interleaving("ab","p","abp");
        assert interleaving("ab","p","apb");
        assert interleaving("ab","p","pab");
        assert interleaving("a","pq","apq");
        assert interleaving("ab","pq","abpq");
        assert interleaving("abc","pqr","abcpqr");
        String s1 = "abcd";
        String s2 = "pqrs";
        String s3 = "apbqcdrs";
        interleaving (s1, s2, s3);
        System.out.println("In main");
    }
}

File: interview/interleaving.cc
// Comparing from Right to left.

#include<string>
#include<iostream>
#include<cassert>

bool interleaving (std::string x, std::string y, std::string z) {
    if (z.length() != x.length() + y.length())
        return false;

    bool A[x.length() + 1][y.length() + 1][z.length() + 1];
    for (int p = 0; p <= x.length() ; p++) 
        for (int q = 0; q <= y.length() ; q++) 
            for (int r = 0 ; r <= z.length(); r++ ) 
                A[p][q][r] = false;

    for (int p = 1; p <= x.length(); p++) 
        if (x.substr(0, p)  == z.substr(0, p))
            A[p][0][p] = true;

    for (int q = 1; q <= y.length(); q++) 
        if (y.substr(0, q) == z.substr(0, q)) 
            A[0][q][q] = true;


    A[0][0][0] = true;


    for (int p = 1; p <= x.length() ; p++) {
        for (int q = 1; q <= y.length() ; q++) {
            for (int r = 1 ; r <= z.length(); r++ ) {
                    int i = p - 1; int j = q - 1; int k = r - 1;
                    if (y[j] == z[k] and x[i] == z[k])
                        A[p][q][r] = A[p][q-1][r-1] or A[p-1][q][r-1];

                    if (y[j] == z[k])
                        A[p][q][r] = A[p][q-1][r-1];

                    else if (x[i] == z[k])
                        A[p][q][r] = A[p-1][q][r-1];
                    else 
                        A[p][q][r] = false;
                    
                    //std::cout << "pqr" << " " << p << " " << q << " " << r << " " ;
                    //std::cout << "x:" << " " << x.substr(0, p) << " " << "y:" << " " << y.substr(0,q)  << " " <<"z:" << " " << z.substr(0,r) << " result " << A[p][q][r] << std::endl ;
            }
        }
    }

    return A[x.length()][y.length()][z.length()];
}

int main (int argc, char** argv) {
    assert(interleaving("","",""));
    assert(interleaving("a","","a"));
    assert(interleaving("b","","b"));
    assert(interleaving("a","p","ap"));
    assert(interleaving("a","p","pa"));
    assert(interleaving("ab","p","abp"));
    assert(interleaving("ab","p","apb"));
    assert(interleaving("ab","p","pab"));
    assert(interleaving("a","pq","apq"));
    assert(interleaving("ab","pq","abpq"));
    assert(interleaving("abc","pqr","abcpqr"));
    assert(not interleaving("abcc","pqr","cabcpqr"));
    assert(not interleaving("aas","pqr","abcpqr"));

    std::string s1 = "abcd";
    std::string s2 = "pqrs";
    std::string s3 = "apbqcdrs";
    assert(interleaving (s1, s2, s3));
}
